File: 20_07_July/0730_42579_베스트앨범/Solution.java
import java.util.ArrayList;
import java.util.HashMap;

class Solution {

    private static class Play {
        private int id;
        private int play;

        public Play(int id, int play) {
            this.id = id;
            this.play = play;
        }
    }

    private static class Genre {
        private String genre;
        private int playTotal = 0;
        private ArrayList<Play> playList = new ArrayList<>();

        public Genre(String genre) {
            this.genre = genre;
        }

        public void addPlay(Play play) {
            this.playTotal += play.play;
            this.playList.add(play);
        }

    }

    public int[] solution(String[] genres, int[] plays) {
        ArrayList<Genre> genreList = new ArrayList<>();
        HashMap<String, Genre> genreMap = new HashMap<>();
        
        // Map을 이용하여 각 데이터 장르별 개체화 후 저장
        for (int i = 0; i < genres.length; i++) {
            String name = genres[i];
            int play = plays[i];
            Genre genre;

            if (genreMap.containsKey(name)) {
                genre = genreMap.get(name);
            } else {
                genre = new Genre(name);

                genreMap.put(name, genre);
                genreList.add(genre);
            }

            genre.addPlay(new Play(i, play));
        }

        // 장르 별 play 총 합계 내림차순
        genreList.sort((a, b) -> Integer.compare(b.playTotal, a.playTotal));

        ArrayList<Integer> results = new ArrayList<>();

        for (Genre g : genreList) {
            // 정렬 기준 : play 내림차순, id 오름차순
            g.playList.sort((a, b) -> (a.play == b.play) ? Integer.compare(a.id, b.id) : Integer.compare(b.play, a.play));
            
            // 최대 2개까지 앨범 결과에 추가
            for (int i = 0; i < Math.min(2, g.playList.size()); i++) {
                results.add(g.playList.get(i).id);
            }
        }

        return results.stream().mapToInt(x -> x.intValue()).toArray();
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        solution.solution(new String[] {"classic", "pop", "classic", "classic", "pop"}, new int[] {500, 600, 150, 800, 2500});
    }

}

File: 20_07_July/0730_42579_베스트앨범/Solution.py

def solution(genres, plays):
    genres_dict = {x: [] for x in genres}
    answer = []

    for id, val in enumerate(zip(genres, plays)):
        genre, play = val[0], val[1]
        genres_dict[genre].append((play, id))

    # 각 장르마다 play 총 합계 저장
    cnt_list = [(sum([x[0] for x in v]), k) for k, v in genres_dict.items()]
    # 앨범 수록곡 총 재생횟수 내림차순 정렬
    cnt_list.sort(key=lambda x: -x[0])
    
    for cnt, genre in cnt_list:
        l = genres_dict[genre]
        # 재생횟수 내림차순, id 오름차순 정렬
        l.sort(key=lambda x: (-x[0], x[1]))
        
        # 최대 2개까지 answer에 추가
        for _, id in l[:min(len(l), 2)]:
            answer.append(id)

    return answer

print(solution(['classic', 'pop', 'classic', 'classic', 'pop'], [500, 600, 150, 800, 2500]))
